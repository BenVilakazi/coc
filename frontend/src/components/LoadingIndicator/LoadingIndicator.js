import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const propType = {
  secondary: PropTypes.bool,
};

const defaultProps = {
  secondary: false,
};

const Loader = styled.div`
  .loading {
    display: inline-block;
    position: relative;
    width: 80px;
    height: 80px;
  }

  .loading div {
    display: inline-block;
    position: absolute;
    left: 8px;
    width: 16px;
    background: ${(props) =>
      props.color
        ? 'var(--primary-background-color)'
        : 'var(--secondary-background-color)'};
    animation: loading 1.2s cubic-bezier(0, 0.5, 0.5, 1) infinite;
  }

  .loading div:nth-child(1) {
    left: 8px;
    animation-delay: -0.24s;
  }

  .loading div:nth-child(2) {
    left: 32px;
    animation-delay: -0.12s;
  }

  .loading div:nth-child(3) {
    left: 56px;
    animation-delay: 0;
  }

  @keyframes loading {
    0% {
      top: 8px;
      height: 64px;
    }
    50%,
    100% {
      top: 24px;
      height: 32px;
    }
  }
`;

export default function LoadingIndicator({ secondary }) {
  return (
    // Need the + to convert the boolean to a number, styled-components requires this as props are passed to the DOM
    // See here: https://github.com/styled-components/styled-components/issues/1198#issuecomment-776416497

    <Loader color={+secondary} data-testid="loader">
      <div className="loading">
        <div />
        <div />
        <div />
      </div>
    </Loader>
  );
}

LoadingIndicator.propTypes = propType;
LoadingIndicator.defaultProps = defaultProps;
