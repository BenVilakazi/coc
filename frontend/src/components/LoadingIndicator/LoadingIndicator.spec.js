import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import LoadingIndicator from './LoadingIndicator';

describe('LoadingIndicator', () => {
  describe('rendering', () => {
    it('renders with the primary-background-color when the "secondary" prop is true', () => {
      render(<LoadingIndicator secondary />);

      expect(screen.getByTestId('loader')).toHaveStyle(
        `background: var(--primary-background-color)`,
      );
    });

    it('renders with the secondary-background-color when the "secondary" prop is false', () => {
      render(<LoadingIndicator secondary={false} />);

      expect(screen.getByTestId('loader')).toHaveStyle(
        `background: var(--secondary-background-color)`,
      );
    });

    it('renders with the secondary-background-color when no prop is passed in', () => {
      render(<LoadingIndicator />);

      expect(screen.getByTestId('loader')).toHaveStyle(
        `background: var(--secondary-background-color)`,
      );
    });

    describe('snapshot', () => {
      it('matches the snapshot', () => {
        const tree = renderer.create(<LoadingIndicator />).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('propTypes', () => {
    it('logs an error if the prop passed in is not a bool', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(<LoadingIndicator secondary="not a bool" />);

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log an error if the prop is left out', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(<LoadingIndicator secondary="not a bool" />);

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
