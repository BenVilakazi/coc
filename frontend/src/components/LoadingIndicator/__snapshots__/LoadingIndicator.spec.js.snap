// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator rendering snapshot matches the snapshot 1`] = `
.c0 .loading {
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
}

.c0 .loading div {
  display: inline-block;
  position: absolute;
  left: 8px;
  width: 16px;
  background: var(--secondary-background-color);
  -webkit-animation: loading 1.2s cubic-bezier(0,0.5,0.5,1) infinite;
  animation: loading 1.2s cubic-bezier(0,0.5,0.5,1) infinite;
}

.c0 .loading div:nth-child(1) {
  left: 8px;
  -webkit-animation-delay: -0.24s;
  animation-delay: -0.24s;
}

.c0 .loading div:nth-child(2) {
  left: 32px;
  -webkit-animation-delay: -0.12s;
  animation-delay: -0.12s;
}

.c0 .loading div:nth-child(3) {
  left: 56px;
  -webkit-animation-delay: 0;
  animation-delay: 0;
}

<div
  className="c0"
  color={0}
  data-testid="loader"
>
  <div
    className="loading"
  >
    <div />
    <div />
    <div />
  </div>
</div>
`;
